### ISSUE TITLE
Move to the Truffle Object Model

### ISSUE THREAD

**Issue by KushalP** [apple/pkl#723]
The [Dynamic Object Model](https://www.graalvm.org/latest/graalvm-as-a-platform/language-implementation-framework/DynamicObjectModel/) APIs were introduced with GraalVM 20.2.0, and the [Static Object Model](https://www.graalvm.org/latest/graalvm-as-a-platform/language-implementation-framework/StaticObjectModel/) APIs were introduced with GraalVM 21.3.0.

Using these object model APIs would be useful for caching, reducing safety checks, and improving profiling.

> [!NOTE]
> Both of the above APIs are not compatible with each other.

**Comment by translatenix** [comment]
Perhaps java.lang.String -> [TruffleString](https://docs.oracle.com/en/graalvm/jdk/21/docs/graalvm-as-a-platform/language-implementation-framework/TruffleStrings/) would also make sense (not sure).

**Comment by translatenix** [comment]
A redesign of the interpreter's object model could explore the following topics:
* Multiple representations for the same type of object
  For example, different representations could be used for small, large, and surrogate listings.
  Truffle supports this via [Truffle Libraries](https://www.graalvm.org/latest/graalvm-as-a-platform/language-implementation-framework/TruffleLibraries/).
* Use of [TruffleString](https://docs.oracle.com/en/graalvm/jdk/21/docs/graalvm-as-a-platform/language-implementation-framework/TruffleStrings/) instead of `java.lang.String`
* Thread safety
   A thread-safe object model could open the door to multi-threaded evaluation of Pkl code in the same `Evaluator`.
   As far as I know, code generated by Truffle is thread-safe.

**Comment by KushalP** [comment]
> Multiple representations for the same type of object

Agreed. As an extension to this it would be interesting to explore if `Listing` and `List` could be consolidated alongside `Mapping` and `Map`. This PR already tries to consolidate the APIs of these two structures, but it could go further with Truffle Libraries to have the "representation" choices around lazy vs. eager be modelled transparently: https://github.com/apple/pkl/pull/683

**Comment by translatenix** [comment]
> As an extension to this it would be interesting to explore if Listing and List could be consolidated alongside Mapping and Map. 

As of 0.26, Listing and List have very different semantics (late binding, etc.) and performance characteristics (lazy amendable object vs. eager persistent collection). Unless that changes, I don't think their internal representations can/should be unified.

**Comment by KushalP** [comment]
> and maybe that's what you'd like to explore?

I'm purely an observer. I'll leave that up to the Pkl team. Mostly want to find ways to help.
